File: test/6_8_goaway.js
var assert = require('assert');

var helper = require('./helper'),
    protocol = helper.protocol;

describe('6.8. GOAWAY', function () {
  it('Sends a GOAWAY frame with the stream identifier that is not 0x0', function (done) {
    var message = 'the endpoint MUST respond with a connection error of type PROTOCOL_ERROR.';

    var conn = helper.createConnection(function () {
      conn.once('error', function (err) {
        assert.equal(err.code, protocol.CODE_PROTOCOL_ERROR, message);
        done();
      });

      var frame = helper.createGoawayFrame(1, protocol.CODE_NO_ERROR, 'h2check');
      frame.streamId = 0x1;
      conn.socket.write(frame.encode());
    });
  });
});

File: client/6_8_goaway.go
package client

import (
	"golang.org/x/net/http2"

	"github.com/summerwind/h2spec/config"
	"github.com/summerwind/h2spec/spec"
)

func GoAway() *spec.ClientTestGroup {
	tg := NewTestGroup("6.8", "GOAWAY")

	// An endpoint MUST treat a GOAWAY frame with a stream identifier
	// other than 0x0 as a connection error (Section 5.4.1) of type
	// PROTOCOL_ERROR.
	tg.AddTestCase(&spec.ClientTestCase{
		Desc:        "Sends a GOAWAY frame with a stream identifier other than 0x0",
		Requirement: "The endpoint MUST treat this as a connection error of type PROTOCOL_ERROR.",
		Run: func(c *config.Config, conn *spec.Conn) error {
			err := conn.Handshake()
			if err != nil {
				return err
			}

			// GOAWAY frame:
			// length: 8, flags: 0x0, stream_id: 1
			conn.Send([]byte("\x00\x00\x08\x07\x00\x00\x00\x00\x01"))
			conn.Send([]byte("\x00\x00\x00\x00\x00\x00\x00\x00"))

			return spec.VerifyConnectionError(conn, http2.ErrCodeProtocol)
		},
	})

	return tg
}
